lib/pbench/test/unit/server/database/__init__.py

```diff
@@ -0,0 +1,222 @@
+        will be reset when the exception is raised.
+        """
+        self.id = 1
+        self.cls = cls
+        self.added: list[Database.Base] = []
+        self.known: dict[int, Database.Base] = {}
+        self.committed: dict[int, FakeRow] = {}
+        self.rolledback = 0
+        self.queries: list[FakeQuery] = []
+        self.raise_on_commit: Optional[Exception] = None
+
+    def query(self, *entities, **kwargs) -> FakeQuery:
+        """
+        Perform a mocked query on the session, setting up the query context
+        and returning it
+
+        Args:
+            entities: The SQLAlchemy entities on which we're operating
+            kwargs: Additional SQLAlchemy parameters
+
+        Returns:
+            A mocked query object
+        """
+        q = FakeQuery(self)
+        self.queries.append(q)
+        return q
+
+    def add(self, instance: Database.Base):
+        """
+        Add a DB object to a list for testing
+
+        Args:
+            instance: A DB object
+        """
+        self.added.append(instance)
+
+    def commit(self):
+        """
+        Mock a commit operation on the DB session. If the 'raise_on_commit'
+        property has been set, "fail" by raising the exception. Otherwise,
+        mock a commit by updating any cached "committed" values if the "known"
+        proxy objects have changed, and record any new "added" objects.
+        """
+        if self.raise_on_commit:
+            exc = self.raise_on_commit
+            self.raise_on_commit = None
+            raise exc
+        for k, object in self.known.items():
+            self.committed[k] = FakeRow.clone(object)
+        for a in self.added:
+            a.id = self.id
+            self.id += 1
+            for c in a.__table__._columns:
+                if c.default:
+                    default = c.default
+                    # print(f"{type(object).__name__}.{c.name}({type(default).__name__}) default:")
+                    # for x in dir(default):
+                    #     print(f"  {x}: {type(getattr(default, x)).__name__} ({getattr(default, x)})")
+                    if default.is_scalar:
+                        setattr(a, c.name, default.arg)
+                    elif default.is_callable:
+                        setattr(a, c.name, default.arg(None))
+            self.known[a.id] = a
+            self.committed[a.id] = FakeRow.clone(a)
+        self.added = []
+
+    def rollback(self):
+        """
+        Just record that rollback was called, since we always raise an error
+        before changing anything during the mocked commit.
+        """
+        self.rolledback += 1
+
+        # Clear the proxy state by removing any new objects that weren't yet
+        # committed, and "rolling back" any proxy values that were updated
+        # from the committed values.
+        self.added = []
+        for k in self.committed:
+            self.known[k] = self.cls(
```

Comment:
Shouldn't we reset the `known` mapping to empty before repopulating it from the committed one?

Reply:
I'm a little fuzzy on what the `known` and `committed` lists constitute.  I thought that `committed` was like a backup of `known`, which we restore on a roll-back...so, when we're rolling back, I expected that we would completely replace the `known` list with a copy of the `committed` list...and the best way to ensure that is to start with an empty list.

However, from the sounds of it, we will be replacing _every item_ in the `known` list, so perhaps there is no point to resetting the mapping to empty (and, there is even an advantage to not doing so, because Python will be able to reuse the data structures which hold the entries).

But, we _will_ be throwing away the `known` IDs that weren't affected by the aborted commit, because we'll be overwriting them with copies of the corresponding entry from the `committed` list.  But, other than being somewhat inefficient, that's fine.